### PR TITLE
Refactor req/resp to better handle different method types

### LINKED ISSUE #808: Refactor req/resp to better handle different method types
The current network request/response code has several hack-ish checks on the Method to handle:
- Methods with no request data
- Methods with no response data
- Methods with only a single response chunk

### PR DESCRIPTION
**Pull request by twoeths**
## Goal
+ resolves #808

### CODE REVIEW

File: packages/lodestar/src/network/reqResp.ts [hunk 1]
```diff
@@ -156,19 +157,19 @@ export class ReqResp extends (EventEmitter as IReqEventEmitterClass) implements
...
+          break;
         }
+        yield* getResponse(peerId, method, data);
       })();
     };
   }
 
   private getResponse = (peerId: PeerId, method: Method, data: Buffer): AsyncIterable<ResponseChunk> => {
-    const request = method === Method.Metadata ? undefined : this.encoder.decodeRequest(method, data);
+    const request = data ? this.encoder.decodeRequest(method, data) : undefined;
```
Comment (mpetrunic) [inline packages/lodestar/src/network/reqResp.ts]: Maybe it's better to handle this inside decodeRequest

File: packages/lodestar/src/network/reqResp.ts [hunk 2]
```diff
@@ -156,19 +157,19 @@ export class ReqResp extends (EventEmitter as IReqEventEmitterClass) implements
...
-        }
+        let data: Buffer;
+        // source is an array of 1 item or empty array
         for await (const val of source) {
-          const data = Buffer.isBuffer(val) ? val : val.slice();
-          yield* getResponse(peerId, method, data);
+          data = Buffer.isBuffer(val) ? val : val.slice();
+          break;
         }
+        yield* getResponse(peerId, method, data);
```
Comment (mpetrunic) [inline packages/lodestar/src/network/reqResp.ts:171]: You could put this before break

File: packages/lodestar/src/constants/network.ts [hunk 3]
```diff
@@ -16,6 +16,11 @@ export enum Method {
   BeaconBlocksByRoot = "beacon_blocks_by_root",
 }
 
+// Methods that returns 1 single response_chunk
+export const SINGLE_CHUNK_METHODS = [Method.Status, Method.Ping, Method.Metadata];
+// Request only method
+export const NO_CHUNK_METHODS = [Method.Goodbye];
```
Comment (mpetrunic) [inline packages/lodestar/src/constants/network.ts]: I don't feel like those arrays are best way forward, as it's quite easy to forget to add it.
We could modify Method enum to contain more informations.
Like:
```javascript
export const Methods = {
   Method.Status: {
      responseType: MethodResponseType.SingleRespone
   },
   Method.Ping: {
      responseType: MethodResponseType.NoResponse
   },
   Method.BlocksByRange: {
      responseType: MethodResponseType.Stream
   }
}
```
Maybe ti makes more sense to add more stuff later like `maxRequestPerSecond` etc

Comment (mpetrunic) [reply packages/lodestar/src/constants/network.ts]: this could also hold ssz types for request and response so we can remove those stuff from encoders

**Review by mpetrunic — APPROVED** [review]

File: packages/lodestar/src/constants/network.ts [hunk 4]
```diff
@@ -16,6 +19,46 @@ export enum Method {
   BeaconBlocksByRoot = "beacon_blocks_by_root",
 }
 
+export enum MethodResponseType {
+  SingleRespone = "SingleRespone",
```
Comment (wemeetagain) [inline packages/lodestar/src/constants/network.ts:23]: Typo

**Review by wemeetagain — DISMISSED** [review]
One typo, otherwise lgtm

File: packages/lodestar/src/constants/network.ts (hunk 4 above)
Comment (mpetrunic) [reply packages/lodestar/src/constants/network.ts:23]: I already fixed that in #841 which is based on this branch so we can merge this

**Pull request merged**